[foreach] Set `SavedVariable.is_output` to `true` for `grad_fn->result_`

crcrpar opened the pull request:
fixes #105502

The scope of this pull request is out-of-place foreach functions that depend on their output tensorlist for backward such as `_foreach_exp`. An example of the generated code with this update is as follows:

```c++
variable_list ForeachExpBackward0::apply(variable_list&& grads) {
  std::lock_guard<std::mutex> lock(mutex_);
  TORCH_CHECK(!result_released_, ERR_BACKWARD_TWICE);
  IndexRangeGenerator gen;
  auto self_ix = gen.range(self_size_);
  variable_list grad_inputs(gen.size());
  auto result = unpack_list(result_, shared_from_this());
  if (task_should_compute_output({ self_ix })) {
    std::vector<Tensor> grad_result;
    grad_result.reserve(grads.size());
    for (const auto & i : c10::irange(grads.size())) {
      if (grads[i].defined()) {
        grad_result.emplace_back(grads[i] * result[i].conj());
      } else {
        grad_result.emplace_back(Tensor());
      }
    }
    copy_range(grad_inputs, self_ix, grad_result);
  }
  return grad_inputs;
}

::std::vector<at::Tensor> _foreach_exp(c10::DispatchKeySet ks, at::TensorList self) {
  auto self_ = unpack(self, "self", 0);
  [[maybe_unused]] auto _any_requires_grad = compute_requires_grad( self );

  std::shared_ptr<ForeachExpBackward0> grad_fn;
  if (_any_requires_grad) {
    grad_fn = std::shared_ptr<ForeachExpBackward0>(new ForeachExpBackward0(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self ));
    grad_fn->self_size_ = self.size();
  }
  #ifndef NDEBUG
  std::vector<c10::optional<Storage>> self__storage_saved(self_.size());
  for (const Tensor& tensor : self_)
    self__storage_saved.push_back(
      tensor.has_storage() ? c10::optional<Storage>(tensor.storage()) : c10::nullopt);
  std::vector<c10::intrusive_ptr<TensorImpl>> self__impl_saved(self_.size());
  for (size_t i=0; i<self_.size(); i++)
    if (self_[i].defined()) self__impl_saved[i] = self_[i].getIntrusivePtr();
  #endif
  auto _tmp = ([&]() {
    if ((isFwGradDefinedTensorList(self))) {
      static c10::OperatorName full_name("aten::_foreach_exp", "");
      static c10::optional<c10::OperatorHandle> opt_op = c10::Dispatcher::singleton().findSchema(full_name);
      return impl::run_jit_decomposition_with_args_for_jvp<::std::vector<at::Tensor>>("_foreach_exp", *opt_op, ks, self);
    } else {
      at::AutoDispatchBelowADInplaceOrView guard;
      return at::redispatch::_foreach_exp(ks & c10::after_autograd_keyset, self_);
    }
  })();
  auto result = std::move(_tmp);
  #ifndef NDEBUG
  for (size_t i=0; i<self_.size() && !at::impl::dispatch_mode_enabled(); i++) {
    if (self__storage_saved[i].has_value() && !at::impl::tensorlist_has_dispatch(self_))
      TORCH_INTERNAL_ASSERT(self__storage_saved[i].value().is_alias_of(self_[i].storage()));
  }
  for (size_t i=0; i<self_.size() && !at::impl::dispatch_mode_enabled(); i++) {
    if (self__impl_saved[i] && !at::impl::tensorlist_has_dispatch(self_))
      TORCH_INTERNAL_ASSERT(self__impl_saved[i] == self_[i].getIntrusivePtr());
  }
  #endif
  if (grad_fn) {
      set_history(flatten_tensor_args( result ), grad_fn);
  }
  if (grad_fn) {
    grad_fn->result_ = make_saved_variable_list(result, true);
  }
  return result;
}
```

A bit of context:
- https://github.com/pytorch/pytorch/pull/105368#issuecomment-1640912479

crcrpar committed:
set `is_output` to `true`

when saving result to `grad_fn`

ref: https://github.com/pytorch/pytorch/issues/105502

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

albanD commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -444,6 +444,13 @@ inline std::vector<SavedVariable> make_saved_variable_list(
       });
 }
 
+inline std::vector<SavedVariable> make_saved_variable_list_foreach_output(
```
nit: you could have added a boolean flag to the existing make_saved_variable_list function.
But sounds ok as well.

albanD reviewed (commented):
Thanks!
Do you have a simple test showing the problem? (a memory leak here).

crcrpar committed:
use TORCH_CHECK, not AT_ASSERT

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
pass `grad_fn` to unpack

```c++
variable_list ForeachExpBackward0::apply(variable_list&& grads) {
  std::lock_guard<std::mutex> lock(mutex_);
  TORCH_CHECK(!result_released_, ERR_BACKWARD_TWICE);
  IndexRangeGenerator gen;
  auto self_ix = gen.range(self_size_);
  variable_list grad_inputs(gen.size());
  auto result = unpack_list(result_, shared_from_this());
  if (task_should_compute_output({ self_ix })) {
    std::vector<Tensor> grad_result;
    grad_result.reserve(grads.size());
    for (const auto & i : c10::irange(grads.size())) {
      if (grads[i].defined()) {
        grad_result.emplace_back(grads[i] * result[i].conj());
      } else {
        grad_result.emplace_back(Tensor());
      }
    }
    copy_range(grad_inputs, self_ix, grad_result);
  }
  return grad_inputs;
}
```

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
merge make_saved_variable_list

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

soulitzer commented on tools/autograd/templates/Functions.h:
```diff
@@ -25,19 +25,19 @@ using at::ScalarType;
   result.reserve(xs.size());
   for (const SavedVariable& v : xs) {
-    auto var = v.unpack();
+    auto var = v.unpack(saved_for);
```
```suggestion
    auto var = v.unpack(std::move(saved_for));
```

soulitzer replied:
Oh whoops, `std::move` does not work because we would be doing it multiple times. 
It really sucks that we need to increment refcount for each input/output, but it would take a larger refactor to do this (no action necessary).

soulitzer commented on tools/autograd/templates/Functions.h:
```diff
@@ -25,19 +25,19 @@ using at::ScalarType;
-  return fmap(xs, [](const SavedVariable& x) {
-    return static_cast<Tensor>(x.unpack());
+  return fmap(xs, [&saved_for](const SavedVariable& x) {
+    return static_cast<Tensor>(x.unpack(saved_for));
```
```suggestion
    return static_cast<Tensor>(x.unpack(std::move(saved_for)));
```

soulitzer reviewed (commented):
Looks great!

I'd add a test like a following show that the grad_fn does not stay alive after the output it corresponds to dies. You can do this by attaching a weakref to the metadata on this grad_fn, deleting it make sure that the weakref is no longer alive:

```python
import weakref
  
def get_ref():
    x = torch.randn(2, 2, requires_grad=True)
    out = torch._foreach_exp([x, x.clone()])

    class Foo:
        pass
    my_obj = Foo()
    meta_dict = out[0].grad_fn.metadata
    meta_dict[0] = my_obj
    ref = weakref.ref(my_obj)
    return out, ref
            
t, ref = get_ref()
self.assertIsNotNone(ref())
del t
self.assertIsNone(ref())
```

crcrpar committed:
`move` `saved_for`

Co-authored-by: soulitzer <soulitzer@gmail.com>

crcrpar commented:
looks like there are a couple of fallouts... one is exposed by the CI, another is I'm seeing with the test based on soulitzer's suggestion.

crcrpar committed:
lifetime test

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
allow for `VectorCType(BaseCType(tensorT))`

as it seems to be the return type of foreach functions with derivatives.yaml entry

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
cover `_foreach_pow.List`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
seems to work w/o move

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented on tools/autograd/gen_autograd_functions.py:
```diff
@@ -583,14 +584,19 @@ def save_var(var: SavedAttribute, is_output: bool) -> None:
+        elif (
+            type == BaseCType(tensorListT)
+            or type == BaseCType(iTensorListRefT)
+            or type == VectorCType(BaseCType(tensorT))
```
At glance this sounded a bit surprising that `Tensor[]` as a return type seems to be treated as `vector<Tensor>` unlike `Tensor[]` being `TensorList` or something equivalent as an argument.

crcrpar replied:
as a context, https://github.com/pytorch/pytorch/blob/a4cffaae675130233fe8a92cf02baad291ec39f9/tools/autograd/derivatives.yaml#L3044-L3046 seems to require this path to allow `vector<Tensor>`

soulitzer replied:
Wait is this specifically for pow? If so, its probably worth a comment.
Maybe worth a comment anyway about arguments and inputs translating into different cpp dtypes.

crcrpar replied:
4fcb0c927ba39d78f974139702b540e3da6d35a5 adding an assert and a comment

soulitzer commented on test/test_foreach.py:
```diff
@@ -913,6 +914,51 @@ def test_outplace_with_invalid_grads(self, device, dtype, op):
+                    sample.input = sample.args[0]
+                    sample.args = new_args
+            _test(func, sample)
+
```
No action necessary, but I'd say that in general, given that this is codegened code, we'd only need a few one-off tests for the representative cases, and testing with all the ops with all the sample inputs might be a little overkill

crcrpar replied:
You're absolutely right. My optimistic expectation here is that the functions in question are unary except for `_foreach_pow`. I also optimistically expect the exception to be lightweight as there should be one tensorlist, multiple scalarlists, and multiple scalars, but all the scalarlists and most scalars would be ignored.

soulitzer commented on test/test_foreach.py:
```diff
@@ -913,6 +914,51 @@ def test_outplace_with_invalid_grads(self, device, dtype, op):
+        ),
+        dtypes=(torch.float32,),
+    )
+    def test_lifetime_of_grad_fn_with_result(self, device, dtype, op):
```
```suggestion
    def test_lifetime_of_grad_fn_when_result_is_saved(self, device, dtype, op):
```

soulitzer reviewed (approved):
Thanks!

crcrpar committed:
rename test to `test_lifetime_of_grad_fn_when_result_is_saved`

Co-authored-by: soulitzer <soulitzer@users.noreply.github.com>
Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
TODO: use `std::move(saved_for)` to avoid bumping refcount

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
note why `VectorCType(BaseCType(tensorT))` for return `Tensor[]`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
@pytorchbot merge

crcrpar commented:
@pytorchbot merge

crcrpar commented:
@pytorchbot merge
